usr/lib/common/loadsave.c

```diff
@@ -239,36 +300,43 @@ CK_RV delete_token_data(STDLL_TokData_t *tokdata)
     return rc;
 }
 
-void init_data_store(STDLL_TokData_t *tokdata, char *directory,
-                     char *data_store)
+CK_RV init_data_store(STDLL_TokData_t *tokdata, char *directory,
+                      char *data_store, size_t len)
 {
     char *pkdir;
+    int splen, pklen;
 
     if (tokdata->pk_dir != NULL) {
         free(tokdata->pk_dir);
         tokdata->pk_dir = NULL;
     }
 
-    if ((pkdir = getenv("PKCS_APP_STORE")) != NULL) {
-        tokdata->pk_dir = (char *) malloc(strlen(pkdir) + 1024);
-        memset(tokdata->pk_dir, 0, strlen(pkdir) + 1024);
-        sprintf(tokdata->pk_dir, "%s/%s", pkdir, SUB_DIR);
-        get_pk_dir(tokdata, data_store);
-        return;
+    if ((pkdir = secure_getenv("PKCS_APP_STORE")) != NULL) {
+        pklen = strlen(pkdir) + 1024;
+        tokdata->pk_dir = (char *) malloc(pklen);
+        memset(tokdata->pk_dir, 0, pklen);
+        splen = snprintf(tokdata->pk_dir, pklen, "%s/%s", pkdir, SUB_DIR);
+        /* Support for old glibc: snprintf might return -1 */
+        if (splen < 0 || splen >= pklen)
+            return -1;
+        return get_pk_dir(tokdata, data_store, len) ? CKR_OK : CKR_FUNCTION_FAILED;
     }
 
     if (directory) {
-        tokdata->pk_dir = (char *) malloc(strlen(directory) + 25);
-        memset(tokdata->pk_dir, 0, strlen(directory) + 25);
-        sprintf(tokdata->pk_dir, "%s", directory);
+        pklen = strlen(directory) + 25;
```

Comment:
Did you find put why we add 25 here?  The s(n)printf a few lines down just fills it with exactly strlen()+1 characters.... So I would understand +1, but why +25?

Reply:
No, I kept the size for now without further investigation on why 25.  I just checked that it is not too small.  Changed this to 1.